tests/stable/testdata/README.md

```diff
@@ -0,0 +1,23 @@
+These stable metrics aren't covered in stablemetrics.yaml, because they aren't
```

Comment:
This is a bit problematic since ideally we would want a system that none of the stable metrics are changed.

Another problem we might encounter is that some labels will not verified because they are not present in the e2e tests. I don't have any example of stable metrics like that in mind, but we have for experimental ones: https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/endpointslice.go#L77-L148.
If we were to verify that metric today in the e2e tests, we would have to make sure that the endpointslice object is complete which is very prone to errors.

I recall we discussed implementing a static analysis solution to make sure that we were going over all the metrics and labels, was that not possible in the end?

Reply:
> a static analysis solution

It's hard to use static analysis for current kube-state-metrics repo, because labels are too dynamic.
For example,  it uses`labelKeys = append(labelKeys, "ready")` in [kube_endpointslice_endpoints](https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/endpointslice.go#L77-L148).

Reply:
1. There are two things that are difficult to handle in that example:
1.1 Having an intermediate labelKeys variable mean that we can't simply have the tool look for the labelKeys field assignation but might need to also build a constant version of the variable assigned to the field. An idea could be to have the tool first look for the labelKeys field assignation and in case the RValue is a variable and not a constant, iterate over the code again to build a const version of that variable.
1.2 `append` should be the only callExp that we will ever see so having a case in the tool just for it would make sense to me. The idea would to replace append(labelKeys, "ready") but building a temporary slice and putting the "ready" string inside of it to convert the variable array to a constant one for our static analysis. 
2. We won't have to handle this scenario because it only affects `_labels` and `_annotations` metrics which have variable labels by design and as such shouldn't have any stability guarantees.

> One possible way to make static analysis easier is to require defining all possible labels first.

This could make sense but I am afraid that this might make label values assignation error-prone since we would have to use the indexes instead of just appending the new labels.

The only option I see today in order to catch all the labels of all the metrics is via static analysis and if we don't want to make that intrusive to the developers to avoid errors, we will need to make the tool intelligent.

For the static analysis I am thinking that the following could be viable:
1. Create an AST of each store and iterate over every generator code
2. Find the labelKeys field assignation
3. If the RValue isn't constant, reiterate over the generators to build a constant version of the variable. That should be possible because even though the slices are dynamic, there values are all constant.

Writing this I noticed that there is another scenario that we would need to handle which is the one of default labels: https://github.com/kubernetes/kube-state-metrics/blob/559bb28d6f35ab797e31cc361605732121a1603e/internal/store/endpointslice.go#L38

These should always be define at the top-level of each store as a constant slice so we should be able to easily add them to each metrics.

This is becoming quite complex so if you have any other suggestions I would gladly hear them but I don't think that we can avoid static analysis if we want to build consistent guarantees.

Reply:
> I don't think that we can avoid static analysis if we want to build consistent guarantees.

Agree

> This could make sense but I am afraid that this might make label values assignation error-prone since we would have to use the indexes instead of just appending the new labels.

Could you provide an example for this case?

Reply:
> Could you provide an example for this case?

I think my reasoning was based on examples such as https://github.com/kubernetes/kube-state-metrics/blob/main/internal/store/endpointslice.go#L77-L148. In these cases order is super important since we build the list of label values one label at a time. I thought that if we were to define label keys upfront, future addition of labels might mess with existing ordering.

But thinking about it again we could transform the code into something like:

```go
const labelKeys []string = {"ready", "serving", ...} 

ready := ""
if ep.Conditions.Ready != nil {
  ready = ep.Conditions.Ready
}

serving := ""
if ep.Conditions.Serving != nil {
  serving = ep.Conditions.Serving
}

m = append(m, &metric.Metric,
  LabelKeys:   labelKeys,
  LabelValues: []string{ready, serving},
  Value:       1,
})
```

This original reasoning for how it is coded in the example I shared above was to reduce the network footprint by removing empty labels. But it could be argued that it doesn't have much impact + we could do processing later on when we build the HTTP response.

So I think your idea of defining const label keys everywhere makes sense and would ease very much static analysis.